File: 섹션08. A형 문제 풀이 (~2022 상반기)/17. 토스트 계란틀 (코드트리, 2018 하반기 오전 2번).cpp
#include <stdio.h>

#define MAX (100 + 20)

int T;

int N, L, R;
int MAP[MAX][MAX];
bool visit[MAX][MAX];

struct RC
{
	int r;
	int c;
};

RC queue[MAX * MAX];

// ↑, →, ↓, ←
int dr[] = { -1, 0, 1, 0 };
int dc[] = { 0, 1, 0, -1 };

void input()
{
	scanf("%d %d %d", &N, &L, &R);

	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= N; c++)
			scanf("%d", &MAP[r][c]);
}

void printMap() // for debug
{
	for (int r = 1; r <= N; r++)
	{
		for (int c = 1; c <= N; c++)
			printf("%2d ", MAP[r][c]);
		putchar('\n');
	}
}

int abs(int x)
{
	return (x > 0) ? x : -x;
}

int BFS(int r, int c)
{
	int rp, wp;
	int sum;

	rp = wp = 0;

	queue[wp].r = r;
	queue[wp++].c = c;

	visit[r][c] = true;

	sum = MAP[r][c];

	while (rp < wp)
	{
		RC out = queue[rp++];

		for (int i = 0; i < 4; i++)
		{
			int nr = out.r + dr[i];
			int nc = out.c + dc[i];

			if (nr < 1 || nc < 1 || nr > N || nc > N) continue;

			int diff = abs(MAP[out.r][out.c] - MAP[nr][nc]);

			if (visit[nr][nc] == true || (L <= diff && diff <= R) == false) continue;

			queue[wp].r = nr;
			queue[wp++].c = nc;

			visit[nr][nc] = true;	
			sum += MAP[nr][nc];
		}
	}

	for (int i = 0; i < wp; i++)
	{
		RC tmp = queue[i];
		MAP[tmp.r][tmp.c] = sum / wp;
	}

	return wp;
}

bool simulate()
{
	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= N; c++)
			visit[r][c] = false;

	bool result = false;
	for (int r = 1; r <= N; r++)
	{
		for (int c = 1; c <= N; c++)
		{
			if (visit[r][c] == true) continue;

			int moveEggCount = BFS(r, c);
			if (moveEggCount != 1) result = true;
		}
	}

	return result;
}

int main()
{
	// scanf("%d", &T);
	T = 1;
	for (int tc = 1; tc <= T; tc++)
	{
		input();

		int answer = 0;

		while (1)
		{
			bool result = simulate();

			if (result == false) break;

			answer++;
		}

		printf("%d\n", answer);
	}

	return 0;
}

File: Python/섹션08. A형 문제 풀이 (~2022 상반기)/17. 토스트 계란틀 (코드트리, 2018 하반기 오전 2번).py
MAX = 100 + 20

MAP = [[0] * MAX for _ in range(MAX)]
visit = [[False] * MAX for _ in range(MAX)]

class RC:
    def __init__(self, r: int, c: int):
        self.r = r
        self.c = c

queue = [RC(0, 0) for _ in range(MAX * MAX)]  

# ↑, →, ↓, ←
dr = [-1, 0, 1, 0]
dc = [0, 1, 0, -1]

def input_data():
    global N, L, R
    
    N, L, R = map(int, input().split())
    
    for r in range(1, N + 1):
        MAP[r][1:N + 1] = map(int, input().split())

def print_map(): # for debug    
    for r in range(1, N + 1):
        print(*MAP[r][1:N + 1])
    print("")

def BFS(r, c):
    rp, wp = 0, 0
    
    queue[wp].r = r
    queue[wp].c = c
    wp += 1
    
    visit[r][c] = True
    
    sum = MAP[r][c]
    
    while rp < wp:
        out = queue[rp]
        rp += 1
        
        for i in range(4):
            nr = out.r + dr[i]
            nc = out.c + dc[i]
            
            if nr < 1 or nc < 1 or nr > N or nc > N: continue
            
            diff = abs(MAP[out.r][out.c] - MAP[nr][nc])
            
            if visit[nr][nc] == True or (L <= diff and diff <= R) == False: continue
            
            queue[wp].r = nr
            queue[wp].c = nc
            wp += 1
            
            visit[nr][nc] = True
            sum += MAP[nr][nc]
    
    for i in range(wp):
        tmp = queue[i]
        MAP[tmp.r][tmp.c] = sum // wp

    return wp

def simulate():
    global visit
    
    visit = [[False] * MAX for _ in range(MAX)]
    
    result = False
    for r in range(1, N + 1):
        for c in range(1, N + 1):
            if visit[r][c] == True: continue
            
            move_egg_count = BFS(r, c)
            if move_egg_count != 1: result = True
     
    return result

# T = int(input())
T = 1
for tc in range(T):
    input_data()
    
    answer = 0
    
    while True:
        result = simulate()
        
        if result == False: break
        
        answer += 1
    
    print(answer)